src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api.Tests/Controllers/DataSetsControllerTests.cs

```diff
@@ -714,440 +719,859 @@ public GetDataSetMetaTests(TestApplicationFactory testApp) : base(testApp)
+                Assert.Equal(
+                    EnumUtil.GetEnums<MetadataType>(),
+                    errorDetail["allowed"].EnumerateArray().Select(e => EnumUtil.GetFromEnumValue<MetadataType>(e.GetString()!)));
+            }
+
+            [Theory]
+            [InlineData("filters", "filters")]
+            [InlineData("locations", "locations")]
+            [InlineData("indicators", "indicators")]
+            [InlineData("timePeriods", "timePeriods")]
+            [InlineData("invalid", "invalid")]
+            [InlineData(null, "")]
+            [InlineData(null, " ")]
+            public async Task InvalidMetadataType_AllowedValueError(string? invalidType, string metadataType)
+            {
+                var response = await GetDataSetMeta(Guid.NewGuid(), types: [metadataType]);
+
+                var validationProblem = response.AssertValidationProblem();
+
+                validationProblem.AssertHasAllowedValueError("types[0]");
+
+                var error = Assert.Single(validationProblem.Errors);
+
+                var errorDetail = GetErrorDetail(error);
+
+                Assert.NotNull(errorDetail);
+                Assert.Equal(2, errorDetail.Count);
+                Assert.Equal(invalidType, errorDetail["value"].GetString());
+                Assert.Equal(
+                    EnumUtil.GetEnums<MetadataType>(),
+                    errorDetail["allowed"].EnumerateArray().Select(e => EnumUtil.GetFromEnumValue<MetadataType>(e.GetString()!)));
+            }
+
+            [Fact]
+            public async Task MultipleInvalidMetadataTypes_AllowedValueError()
+            {
+                var response = await GetDataSetMeta(Guid.NewGuid(), types: ["invalid1", "invalid2"]);
+
+                var validationProblem = response.AssertValidationProblem();
+
+                validationProblem.AssertHasAllowedValueError("types[0]");
+                validationProblem.AssertHasAllowedValueError("types[1]");
+
+                Assert.Equal(2, validationProblem.Errors.Count);
+
+                var errorDetail1 = GetErrorDetail(validationProblem.Errors[0]);
+                var errorDetail2 = GetErrorDetail(validationProblem.Errors[1]);
+
+                Assert.NotNull(errorDetail1);
+                Assert.NotNull(errorDetail2);
+                Assert.Equal(2, errorDetail1.Count);
+                Assert.Equal(2, errorDetail2.Count);
+                Assert.Equal("invalid1", errorDetail1["value"].GetString());
+                Assert.Equal("invalid2", errorDetail2["value"].GetString());
+                Assert.Equal(
+                    EnumUtil.GetEnums<MetadataType>(),
+                    errorDetail1["allowed"].EnumerateArray().Select(e => EnumUtil.GetFromEnumValue<MetadataType>(e.GetString()!)));
+                Assert.Equal(
+                    EnumUtil.GetEnums<MetadataType>(),
+                    errorDetail2["allowed"].EnumerateArray().Select(e => EnumUtil.GetFromEnumValue<MetadataType>(e.GetString()!)));
```

Comment:
Is there potentially a nicer way for us to do this? We could a lot of this logic into reusable extension on `ValidationProblem`.

```cs
validationProblem.AssertHasAllowedValueError(
    path:"types[0]",
    value: "invalid1",
    allowed: EnumUtil.GetEnumValues<MetadataType>());

validationProblem.AssertHasAllowedValueError(
    path:"types[1]",
    value: "invalid1",
    allowed: EnumUtil.GetEnumValues<MetadataType>());
```

Reply:
I've changed the extension method to do these extra bits for us :)